### PR TITLE
Add github_webhook_facts module

### PR DESCRIPTION
**Pull request by stpierre**
##### SUMMARY

`github_webhook_facts` gathers information about Github webhooks.

This is part of a two-part replacement for `github_hooks`, which lacks a number of important features and has a strange interface.

The other parts are:

* Add `github_webhook` module: https://github.com/ansible/ansible/pull/35813
* Deprecate `github_hooks` module: #49296

`github_hooks` lacks a number of important features: The ability to specify the events on which a webhook should fire; the ability to add a shared secret to a hook; the ability to create hooks that do not verify SSL against their target; and more. I considered merely updating that module to add the features needed, but the existing interface is pretty odd and would be difficult to update in a backwards-compatible way. For instance, it has a built-in feature for deleting all hooks whose last request returned 504 (`action: clean504`), but no way to update an existing hook, or delete a specific hook. It requires you to provide the API URL to a repository, rather than just the name of the repo (and optionally a Github API base), which requires a fair degree of familiarity with the Github API.

It has other issues that are easier to fix, but still weighed in on my decision to replace rather than upgrade: it rolls its own API client layer rather than using an existing library; it appears to support both password and API token auth, but only accepts a single `oauthkey` argument, which is a misnomer in both cases.

Rather than try to update it and fix all of the issues, I replaced it with two modules whose usage patterns are, I hope, much more intuitive.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME
github_webhook_facts

##### ANSIBLE VERSION
```
ansible 2.5.0 (github-webhook-facts-module 0192267b05) last updated 2018/02/06 16:11:06 (GMT -500)
  config file = None
  configured module search path = [u'/home/stpierre/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/stpierre/devel/ansible/lib/ansible
  executable location = /home/stpierre/devel/ansible/bin/ansible
  python version = 2.7.14 (default, Jan 17 2018, 14:28:32) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]
```

### CODE REVIEW

**Comment by djmattyg007** [comment]
shipit

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_webhook_facts.py]: s/Github/GitHub/

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_webhook_facts.py]:
```
# Copyright: (c) 2018, Ansible Project
```

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_webhook_facts.py:28]:
```
aliases: ['repo']
```

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_webhook_facts.py:47]: Does github_url will be different for enterprise use ? If it is going to be common for all then no need to provide this option.

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_webhook_facts.py:126]: ~I do not think so `check mode` is required for facts module.~

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_webhook_facts.py:23]: Could you please use `github3.py` instead of `PyGithub` as this will match with other github_* modules ?

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_webhook_facts.py:156]: check if user is authenticated with given password or token ?

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (Akasurde) [inline lib/ansible/modules/source_control/github_webhook_facts.py]:
```
"Unable to get hooks from repository %s: %s" % to_native(err)
```
to 
```
Unable to get hooks from repository :%s" % to_native(err)
```

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (stpierre) [reply lib/ansible/modules/source_control/github_webhook_facts.py:47]: Yes, it will be different for GitHub Enterprise.

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (stpierre) [reply lib/ansible/modules/source_control/github_webhook_facts.py:23]: As I mentioned on the other review, `github3.py` has pretty bad GitHub Enterprise support, so no.

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (djmattyg007) [reply lib/ansible/modules/source_control/github_webhook_facts.py:126]: What does this mean? Why would you not want facts during a check mode run?

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (Akasurde) [reply lib/ansible/modules/source_control/github_webhook_facts.py:23]: I think this will add additional overhead of separate library to both user and maintainer. So -1 from my side.

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (Akasurde) [reply lib/ansible/modules/source_control/github_webhook_facts.py:126]: ~This flag denotes that module will not perform any changes to remote system if `--check` is provided while running. Fact modules does not change anything on remote systems so it is pointless to have `supports_check_mode` as True. Moreover, there is no action performed with respect to `module.check_mode` like done [here](https://github.com/ansible/ansible/blob/2ad7d79985fef9abc2ee1c5166762b5773cf5c4d/lib/ansible/modules/cloud/vmware/vmware_guest.py#L2095).~

**Review by Akasurde — DISMISSED** [review]

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (stpierre) [reply lib/ansible/modules/source_control/github_webhook_facts.py:23]: So what would be an appropriate path forward here? `github3.py` is really bad at talking to GitHub Enterprise, so it's clearly not an option. Are you suggesting that I first port the other GitHub-related models to PyGithub in order to merge this?

Blocking this review on the fact that the *other* Github-related modules made a poor choice of library isn't terribly reasonable IMO. So I'd like to know what you suggest to move forward. (Aside from using `github3.py`, of course, since it doesn't work.)

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (stpierre) [reply lib/ansible/modules/source_control/github_webhook_facts.py:126]: If a module does not support check mode (i.e., `supports_check_mode=False`), then it will be skipped when run in check mode. If we skip a facts module in check mode, then it could significantly impact the run -- facts could be missing that many tasks depend on, for instance. Additionally, since this module *never* changes anything, it doesn't need to do anything special for check mode.

So yes, this module *does* support check mode: it always runs in a manner that doesn't change anything, but only predicts if anything would be changed (spoiler alert: it wouldn't). And moreover, changing this line to `False` could have unnecessary negative effects if someone were to depend on facts from this module that would then become unavailable in check mode _for no good reason_.

Comment (stpierre) [reply lib/ansible/modules/source_control/github_webhook_facts.py:126]: To put it differently:

`supports_check_mode=True` does *not* mean "this module does something differently in check mode." It *does* mean "this module can be run in a way that doesn't make changes, only reports them." And that's what this module does *every* time, so clearly it supports check mode.

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (Akasurde) [reply lib/ansible/modules/source_control/github_webhook_facts.py:23]: You are taking me wrong here. I am just saying that adding additional library will add overhead to people who are using this module and people who will maintain it. I am not saying to port other GitHub modules to use `PyGithub`. One way could be using GitHub REST APIs directly rather than relying on any third party Python modules.

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (Akasurde) [reply lib/ansible/modules/source_control/github_webhook_facts.py:126]: Make sense to me. Thanks for info.

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (stpierre) [reply lib/ansible/modules/source_control/github_webhook_facts.py:23]: I don't see how that's a good choice. It seems to me that the other modules made a poor choice of library, and so going forward we should choose a better one. I've selected PyGithub, but I'm obviously open to other suggestions. What I'm not open to is either a) being required to use the same broken library as the other modules; or b) reinventing the wheel.

Comment (gundalow) [reply lib/ansible/modules/source_control/github_webhook_facts.py:23]: @adrianmoisey @Akasurde @andreparames @bincyber @dj-wasabi @djmattyg007 @dsummersl @pcgentry @yeukhon

I believe you've worked on other GitHub modules, it would be good to get your views on this

Comment (Akasurde) [reply lib/ansible/modules/source_control/github_webhook_facts.py:23]: I would say then port rest of the github modules using `PyGithub` library so that everyone is benefited. What say ?

Comment (stpierre) [reply lib/ansible/modules/source_control/github_webhook_facts.py:23]: IMO that sets an unreasonably high barrier to entry, and it's something I have neither the time nor desire to do ATM. I'll leave this PR here in case someone else wants to do it in the future, but it's going to be more appropriate for us to just maintain our own version of this module if that's the cost of getting it merged.

Comment (djmattyg007) [reply lib/ansible/modules/source_control/github_webhook_facts.py:23]: @Akasurde I think you're being unreasonable. There is no technical reason for rejecting this patch. The dependencies for each module are clearly indicated in documentation. I don't see how that could be considered confusing. IMO there are far more pressing quality concerns for Ansible's suite of modules than which API wrapper each github module uses.

Comment (Akasurde) [reply lib/ansible/modules/source_control/github_webhook_facts.py:23]: OK I am dismissing my all reviews. Thanks.

Comment (gundalow) [reply lib/ansible/modules/source_control/github_webhook_facts.py:23]: Making this as resolved.

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (gundalow) [inline lib/ansible/modules/source_control/github_webhook_facts.py]:
```suggestion
- name: list hooks for a repository (password auth)
  github_webhook_facts:
```

File: lib/ansible/modules/source_control/github_webhook_facts.py
Comment (gundalow) [inline lib/ansible/modules/source_control/github_webhook_facts.py]:
```suggestion
- name: list hooks for a repository on GitHub enterprise (token auth)
  github_webhook_facts:
```

**Review by gundalow — CHANGES_REQUESTED** [review]

**Review by gundalow — APPROVED** [review]
Made the changes myself, I'll merge once CI is green

**Comment by gundalow** [comment]
@stpierre Thanks, got there eventually, merged into `devel` for release in Ansible 2.8

**Pull request merged**
